### DISCUSSION TITLE
Questions regarding dns_lookup_int() in lib/roken/resolve.c

### DISCUSSION THREAD

**Discussion by NormProffitt** [heimdal/heimdal#1198] [General]
I'm looking through the `dns_lookup_int()` function in `lib/roken/resolve.c` and have some questions.

Once in the `while(1)` loop I see the call to the `#defined resolve_search()` function followed by the checking of its return value.  I understand the `if (size > len)` and the `if (size > 0)` checks but I'm confused by what's done in the `if (size <= 0 && len < rk_DNS_MAX_PACKET_SIZE)` test.

First, when would the `resolve_search()` function (either `dns_search()`, `res_nsearch()`, or `res_search()` depending on the `#ifdefined`) return 0?  From the documentation I've found those underlying functions either return the size of the DNS response or -1 if there's an error.  Even if the DNS resource record is not found I would expect some sort of reply with a DNS response code like `NXDOMAIN`, etc. and not a zero byte reply.

Also, even if the resolver returned a zero byte response or -1 for an error why would doubling the size of the reply buffer hope to improve things?  For example, `NXDOMAIN` is `NXDOMAIN` and `ETIMEDOUT` is `ETIMEDOUT` no matter how big a reply buffer you provide. Doubling the reply buffer size and retrying multiple times can significantly delay this function returning an error if the reason for the error is `ETIMEDOUT`.

Any insight into why the `if (size <= 0 && len < rk_DNS_MAX_PACKET_SIZE)` check is doing what it's doing would be greatly appreciated.

**Comment by NormProffitt** [comment]
Is this the correct place for me to be asking these kind of questions?

**Reply by nicowilliams** [reply]
Yes, it is.  Sorry we've been slow to respond.

**Comment by nicowilliams** [comment]
> Also, even if the resolver returned a zero byte response or -1 for an error why would doubling the size of the reply buffer hope to improve things? For example, `NXDOMAIN` is `NXDOMAIN` and `ETIMEDOUT` is `ETIMEDOUT` no matter how big a reply buffer you provide. Doubling the reply buffer size and retrying multiple times can significantly delay this function returning an error if the reason for the error is `ETIMEDOUT`.
> 
> Any insight into why the `if (size <= 0 && len < rk_DNS_MAX_PACKET_SIZE)` check is doing what it's doing would be greatly appreciated.

The resolver functions return the size of the reply irrespective of whether the reply fit in the provided buffer.  The loop is about resizing the buffer until it fits the reply.

From this code I suspect that there are systems where the resolver functions return `-1` to indicate that the provided buffer was too small.  If there are no systems remaining where this is the case that we want to support we could optimize this away and fail faster (due to timeouts).
